File: TP 3/verificar_ingreso.c
#include "verificar_ingreso.h"
#include "heroe.h"
#include "comandos.h"
#include <stdlib.h>

#define MAX_COMANDO 15

const int ARGS_LISTAR = 3;
const int ARGS_CONTACTAR = 4;
const int ARGS_MODIFICAR = 6;
const int ARGS_AGREGAR = 7;
const int ARGS_AYUDA = 2;
const int EDAD_MINIMA = 1;
const int EDAD_MAXIMA = 100;
const int ID_MINIMA = 1;

//PRE: -
//POS: Retorna true si es la edad esta entre 1 y 100.
bool edad_valida(int edad){
    bool valido = true;
    if(edad < EDAD_MINIMA || edad > EDAD_MAXIMA){
        printf("La edad ingresada es inválida, pruebe el comando ayuda\n");
        valido = false;
    }
    return valido;
}

//PRE: -
//POS: Retorna true si el nombre es menor a 50 caracteres.
bool nombre_valido(char nombre[MAX_NOMBRE]){
    bool valido = true;
    if(strlen(nombre) > MAX_NOMBRE){
        printf("El nombre ingresado es muy largo, pruebe el comando ayuda\n");
        valido = false;
    }
    return valido;
}

//PRE: -
//POS: Retorna true si el estado es vivo o muerto.
bool estado_valido(char estado[MAX_ESTADO]){
    bool valido = true;
    if((strcmp(estado,VIVO_COMPLETO)!=0) && strcmp(estado,MUERTO_COMPLETO)!=0){
        printf("El estado ingresado es erróneo, pruebe el comando ayuda\n");
        valido = false;
    }
    return valido;
}

//PRE: -
//POS: Retorna true si la id es un entero positivo.
bool id_valido(int id){
    bool valido = true;
    if(id < ID_MINIMA){
        printf("La ID ingresada es errónea, pruebe el comando ayuda\n");
        valido = false;
    }
    return valido;
}

//PRE: -
//POS: Devuelve true si son una cantidad correcta de argumentos.
bool cant_argumentos_valida(int cant_args){
    bool valido = true;
    if(cant_args != ARGS_LISTAR && cant_args != ARGS_CONTACTAR && cant_args != ARGS_MODIFICAR &&
     cant_args != ARGS_AGREGAR && cant_args != ARGS_AYUDA){
        printf("La cantidad de argumentos ingresada es inválida, pruebe el comando ayuda\n");
        valido = false;
    }
    return valido;
}

//PRE: -
//POS: Devuelve true si es un comando existente.
bool comando_valido(char comando[MAX_COMANDO]){
    bool valido = true;
    if((strcmp(comando,LISTAR) != 0) && (strcmp(comando,AGREGAR) != 0) && (strcmp(comando,CONTACTAR) != 0) && 
    strcmp(comando,MODIFICAR) != 0 && (strcmp(comando,AYUDA) != 0)){
        printf("Nombre del comando inválido, pruebe el comando ayuda\n");
        valido = false;
    }
    return valido;
}

//PRE: -
//POS: Devuelve true si el archivo tiene extension .csv.
bool ruta_valida(char ruta[MAX_RUTA]){
    bool valido = true;
    if(strstr(ruta,".csv") == NULL){
        printf("La ruta ingresada no contiene la extension .csv, pruebe el comando ayuda\n");
        valido = false;
    }
    return valido;
}

int ingresos_validos(int argc, char* argv[]){

    int codigo = ERROR_INGRESO_INVALIDO;

    char comando[MAX_COMANDO];
    strcpy(comando, argv[1]);

    if((strcmp(comando,AYUDA) == 0) && argc == ARGS_AYUDA){
        return NUMERO_AYUDA;
    }

    if(comando_valido(comando) && cant_argumentos_valida(argc) && ruta_valida(argv[argc -1])){
        
        if((strcmp(comando,LISTAR) == 0) && argc == ARGS_LISTAR){
            codigo = NO_HAY_ERROR;
        }
        else if((strcmp(comando,AGREGAR) == 0) && argc == ARGS_AGREGAR && id_valido(atoi(argv[2])) &&
        nombre_valido(argv[3]) && edad_valida(atoi(argv[4])) && estado_valido(argv[5])){
                codigo = NO_HAY_ERROR;
        }
        else if((strcmp(comando,MODIFICAR) == 0) && argc == ARGS_MODIFICAR && id_valido(atoi(argv[2])) &&
         edad_valida(atoi(argv[3])) && estado_valido(argv[4])){
                codigo = NO_HAY_ERROR;
        }
        else if((strcmp(comando,CONTACTAR) == 0) && argc == ARGS_CONTACTAR && id_valido(atoi(argv[2]))){
                codigo = NO_HAY_ERROR;
        } 
    }

    return codigo;
}

File: TP3 - PYTHON/verificar_ingreso.py
import codigo_error as code
import comandos as cmd
import heroe

ARGS_LISTAR = 3
ARGS_AYUDA = 2
ARGS_CONTACTAR = 4
ARGS_MODIFICAR = 6
ARGS_AGREGAR = 7
ARGS_AYUDA = 2
EDAD_MINIMA = 1
EDAD_MAXIMA = 100
ID_MINIMA = 1

#PRE: -
#POS: Devuelve true si la edad es un entero entre 1 y 100.
def edad_valida(edad):
    valido = True
    if(int(edad) < EDAD_MINIMA or int(edad) > EDAD_MAXIMA):
        print("La edad ingresada es inválida, pruebe el comando ayuda")
        valido = False
    return valido

#PRE: -
#POS:Retorna true si el nombre es menor a 50 caracteres.
def nombre_valido(nombre):
    valido = True
    if(len(nombre) > heroe.MAX_NOMBRE):
        print("El nombre ingresado es inválido, pruebe el comando ayuda")
        valido = False
    return valido

#PRE: -
#POS: Retorna true si el estado es vivo o muerto.
def estado_valido(estado):
    valido = True
    if(estado != heroe.MUERTO_COMPLETO and estado != heroe.VIVO_COMPLETO):
        print("El estado ingresado es inválido, pruebe el comando ayuda")
        valido = False
    return valido

#PRE: -
#POS: Retorna true si la id es un entero positivo.
def id_valida(id):
    valido = True
    if(int(id) < ID_MINIMA):
        print("La id ingresada es inválida, pruebe el comando ayuda")
        valido = False
    return valido

#PRE: -
#POS: Devuelve true si son una cantidad correcta de argumentos.
def cant_argumentos_valida(cant_args):
    valido = True
    if(cant_args != ARGS_LISTAR and cant_args != ARGS_MODIFICAR and cant_args != ARGS_AGREGAR and cant_args != ARGS_AYUDA and cant_args != ARGS_CONTACTAR):
        print("La cantidad de argumentos ingresada es inválida, pruebe el comando ayuda")
        valido = False
    return valido

#PRE: -
#POS: Devuelve true si es un comando existente.
def comando_valido(comando):
    valido = True
    if(comando != cmd.AGREGAR and comando != cmd.MODIFICAR and comando != cmd.LISTAR and 
    comando != cmd.AYUDA and comando != cmd.CONTACTAR):
        print("Nombre del comando inválido, pruebe el comando ayuda")
        valido = False
    return valido

#PRE: -
#POS: Devuelve true si el archivo tiene extension .csv.
def ruta_valida(ruta):
    valido = True
    if not(".csv" in ruta):
       print("La ruta ingresada no contiene la extension .csv, pruebe el comando ayuda")
       valido = False  

    return valido

#PRE: -
#POS: Devuelve 0 si todo se ingresó correctamente. 1 si el comando es ayuda. Sino -7.
def ingresos_validos(argc, argv):
    codigo = code.ERROR_INGRESO_INVALIDO

    comando = argv[1]

    if(comando == cmd.AYUDA and argc == ARGS_AYUDA):
        return code.NUMERO_AYUDA
    
    if (comando_valido(comando) and cant_argumentos_valida(argc) and ruta_valida(argv[argc -1])):
        if(comando == cmd.LISTAR and argc == ARGS_LISTAR):
            codigo = code.NO_HAY_ERROR
        elif(comando == cmd.MODIFICAR and argc == ARGS_MODIFICAR and id_valida(argv[2]) and
        edad_valida(argv[3]) and estado_valido(argv[4])):
            codigo = code.NO_HAY_ERROR
        elif(comando == cmd.AGREGAR and argc == ARGS_AGREGAR and id_valida(argv[2]) and 
        nombre_valido(argv[3]) and edad_valida(argv[4]) and estado_valido(argv[5])):
            codigo = code.NO_HAY_ERROR
        elif(comando == cmd.CONTACTAR and argc == ARGS_CONTACTAR and id_valida(argv[2])):
            codigo = code.NO_HAY_ERROR
            
    return codigo